core/src/main/kotlin/users/service/UserService.kt

```diff
@@ -195,4 +232,120 @@ class UserServiceImpl(
+            "<b>${user.email}로 가입된 아이디는 다음과 같습니다.</b><br/><br/>" +
+            "<h3>아이디</h3><h3>${user.credential.localId}</h3><br/><br/>"
+        mailClient.sendMail(email, emailSubject, emailBody)
+    }
+
+    override suspend fun sendResetPasswordCode(email: String) {
+        if (!authService.isValidEmail(email)) throw InvalidEmailException
+        val user = userRepository.findByEmailAndActiveTrue(email) ?: throw UserNotFoundException
+        val key = "reset-password-code-${user.id}"
+        val code = Base64.getUrlEncoder().encodeToString(Random.nextBytes(6))
+        val value = getVerificationValue(email, code, key)
+        val emailSubject = "[SNUTT] 인증코드 [$code] 를 입력해주세요"
+        val emailBody = "<h2>비밀번호 재설정 안내</h2><br/>" +
+            "안녕하세요. SNUTT입니다. <br/> " +
+            "<b>아래의 인증코드를 진행 중인 화면에 입력하여 비밀번호 재설정을 완료해주세요.</b><br/><br/>" +
+            "<h3>인증코드</h3><h3>$code</h3><br/><br/>" +
+            "인증코드는 이메일 발송 시점으로부터 3분 동안 유효합니다."
+        redisTemplate.opsForValue().set(key, mapper.writeValueAsString(value), Duration.ofMinutes(3)).subscribe()
+        mailClient.sendMail(email, emailSubject, emailBody)
+    }
+
+    override suspend fun verifyResetPasswordCode(localId: String, code: String) {
+        val user = userRepository.findByCredentialLocalIdAndActiveTrue(localId) ?: throw UserNotFoundException
+        val key = "reset-password-code-${user.id}"
+        val value = mapper.treeToValue(mapper.reader().readTree(redisTemplate.opsForValue().getAndAwait(key)), RedisVerificationValue::class.java)
+        if (value == null || value.code != code) throw InvalidVerificationCodeException
+        redisTemplate.opsForValue().getAndExpire(key, Duration.ofMinutes(3)).subscribe()
+    }
+
+    override suspend fun getMaskedEmail(localId: String): String {
+        val user = userRepository.findByCredentialLocalIdAndActiveTrue(localId) ?: throw UserNotFoundException
+        val email = user.email ?: throw UserNotFoundException
+        val maskedEmail = email.replace(Regex("(?<=.{3}).(?=.*@)"), "*")
+        return maskedEmail
+    }
+
+    override suspend fun resetPassword(localId: String, newPassword: String, code: String) {
+        val user = userRepository.findByCredentialLocalIdAndActiveTrue(localId) ?: throw UserNotFoundException
+        verifyResetPasswordCode(localId, code)
+        if (!authService.isValidPassword(newPassword)) throw InvalidPasswordException
+        user.credential.localPw = authService.buildLocalCredential(user.credential.localId!!, newPassword).localPw
+        user.credentialHash = authService.generateCredentialHash(user.credential)
+        userRepository.save(user)
+        redisTemplate.delete("reset-password-code-${user.id}").subscribe()
+    }
+
+    private suspend fun getVerificationValue(email: String, code: String, key: String): RedisVerificationValue {
+        val existing = when (val existingRedisString = redisTemplate.opsForValue().getAndAwait(key)) {
```

Comment:
서버 만들때 코드라인이 길어지는걸 피하지 말아주세요
```suggestion
        val existingRedisString = redisTemplate.opsForValue().getAndAwait(key)
        val existing = when (existingRedisString) {
```